In [1]:
# Designating one cell for all imports

import pandas as pd

In [2]:
!ls

Initial_EDA_beeradvocate.ipynb


In [3]:
df_ba = pd.read_csv('../../data/beer_reviews.csv')


#df_ba stands for initial beer advocate data

In [4]:
df_ba.head()

brewery_id             brewery_name  review_time  review_overall  \
0       10325          Vecchio Birraio   1234817823             1.5   
1       10325          Vecchio Birraio   1235915097             3.0   
2       10325          Vecchio Birraio   1235916604             3.0   
3       10325          Vecchio Birraio   1234725145             3.0   
4        1075  Caldera Brewing Company   1293735206             4.0   

   review_aroma  review_appearance review_profilename  \
0           2.0                2.5            stcules   
1           2.5                3.0            stcules   
2           2.5                3.0            stcules   
3           3.0                3.5            stcules   
4           4.5                4.0     johnmichaelsen   

                       beer_style  review_palate  review_taste  \
0                      Hefeweizen            1.5           1.5   
1              English Strong Ale            3.0           3.0   
2          Foreign / Export Stout            3.0           3.0   
3                 German Pilsener            2.5           3.0   
4  American Double / Imperial IPA            4.0           4.5   

                beer_name  beer_abv  beer_beerid  
0            Sausa Weizen       5.0        47986  
1                Red Moon       6.2        48213  
2  Black Horse Black Beer       6.5        48215  
3              Sausa Pils       5.0        47969  
4           Cauldron DIPA       7.7        64883

In [5]:
df_ba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   brewery_id          1586614 non-null  int64  
 1   brewery_name        1586599 non-null  object 
 2   review_time         1586614 non-null  int64  
 3   review_overall      1586614 non-null  float64
 4   review_aroma        1586614 non-null  float64
 5   review_appearance   1586614 non-null  float64
 6   review_profilename  1586266 non-null  object 
 7   beer_style          1586614 non-null  object 
 8   review_palate       1586614 non-null  float64
 9   review_taste        1586614 non-null  float64
 10  beer_name           1586614 non-null  object 
 11  beer_abv            1518829 non-null  float64
 12  beer_beerid         1586614 non-null  int64  
dtypes: float64(6), int64(3), object(4)
memory usage: 157.4+ MB


In [6]:
df_ba.isna().sum()

brewery_id                0
brewery_name             15
review_time               0
review_overall            0
review_aroma              0
review_appearance         0
review_profilename      348
beer_style                0
review_palate             0
review_taste              0
beer_name                 0
beer_abv              67785
beer_beerid               0
dtype: int64

Looks like we have some missing values to deal with:

1. 15 in brewery_name
2. 348 in review_profilename
3. ~67k in beer_abv

## Dealing with brewery_name

In [7]:
df_ba[df_ba['brewery_name'].isna()]

brewery_id brewery_name  review_time  review_overall  review_aroma  \
651565         1193          NaN   1301022066             2.0           2.5   
659293         1193          NaN   1290107698             4.0           4.5   
659299         1193          NaN   1289077001             3.5           3.0   
659300         1193          NaN   1289851033             3.5           4.0   
659301         1193          NaN   1289158632             3.5           4.0   
659302         1193          NaN   1289590065             3.5           4.0   
659303         1193          NaN   1298078926             3.0           3.0   
659304         1193          NaN   1292423271             4.5           4.0   
659305         1193          NaN   1289329962             3.5           3.5   
1391043          27          NaN   1020244856             3.5           3.5   
1391049          27          NaN   1143182956             2.5           2.0   
1391050          27          NaN   1117963232             1.5           1.5   
1391051          27          NaN   1323314674             4.5           4.5   
1391052          27          NaN   1320989774             3.5           4.0   
1391053          27          NaN   1320995408             5.0           4.5   

         review_appearance review_profilename                  beer_style  \
651565                 2.5            Knapp85                Vienna Lager   
659293                 3.5             dqrull                        Bock   
659299                 3.0             dqrull   Dortmunder / Export Lager   
659300                 3.5             dqrull         Munich Helles Lager   
659301                 4.0             dqrull  Keller Bier / Zwickel Bier   
659302                 3.0             dqrull                Vienna Lager   
659303                 3.0         Ochsenblut  Keller Bier / Zwickel Bier   
659304                 4.0         Dentist666  Keller Bier / Zwickel Bier   
659305                 4.0             dqrull  Keller Bier / Zwickel Bier   
1391043                3.5              Jason      American Adjunct Lager   
1391049                2.0            BeerBob        American Malt Liquor   
1391050                2.0      feloniousmonk        American Malt Liquor   
1391051                3.5             ccrida                American IPA   
1391052                3.5              Docer                American IPA   
1391053                4.0              Docer              American Stout   

         review_palate  review_taste  \
651565             1.5           1.5   
659293             4.0           3.5   
659299             4.0           4.0   
659300             3.5           3.0   
659301             4.0           3.5   
659302             3.0           3.5   
659303             2.0           3.0   
659304             4.0           4.5   
659305             4.0           4.0   
1391043            4.0           3.5   
1391049            2.0           2.0   
1391050            1.5           1.5   
1391051            4.5           4.5   
1391052            4.0           4.0   
1391053            4.5           4.5   

                                                 beer_name  beer_abv  \
651565   Engel Tyrolian Bräu WRONG BREWERY SEE SCHWABIS...       5.0   
659293    Engel Bock Dunkel WRONG BREWERY SEE CRAILSHEIMER       7.2   
659299           Engel Gold WRONG BREWERY SEE CRAILSHEIMER       5.4   
659300       Engel Landbier WRONG BREWERY SEE CRAILSHEIMER       4.8   
659301    Engel Keller Hell WRONG BREWERY SEE CRAILSHEIMER       5.4   
659302     Engel Aloisius - WRONG BREWERY SEE CRAILSHEIMER       5.9   
659303   Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659304   Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659305   Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
1391043                             Hard Hat American Beer       3.8   
1391049                                        Side Pocket       8.5   
139105

Just looks like we're missing brewery_name for brewery_id that represent 1193 and 27. Let's take a look to see if we have values for these in other data entries.

In [8]:
df_ba[df_ba['brewery_id'] == 1193]

brewery_id brewery_name  review_time  review_overall  review_aroma  \
651565        1193          NaN   1301022066             2.0           2.5   
659293        1193          NaN   1290107698             4.0           4.5   
659299        1193          NaN   1289077001             3.5           3.0   
659300        1193          NaN   1289851033             3.5           4.0   
659301        1193          NaN   1289158632             3.5           4.0   
659302        1193          NaN   1289590065             3.5           4.0   
659303        1193          NaN   1298078926             3.0           3.0   
659304        1193          NaN   1292423271             4.5           4.0   
659305        1193          NaN   1289329962             3.5           3.5   

        review_appearance review_profilename                  beer_style  \
651565                2.5            Knapp85                Vienna Lager   
659293                3.5             dqrull                        Bock   
659299                3.0             dqrull   Dortmunder / Export Lager   
659300                3.5             dqrull         Munich Helles Lager   
659301                4.0             dqrull  Keller Bier / Zwickel Bier   
659302                3.0             dqrull                Vienna Lager   
659303                3.0         Ochsenblut  Keller Bier / Zwickel Bier   
659304                4.0         Dentist666  Keller Bier / Zwickel Bier   
659305                4.0             dqrull  Keller Bier / Zwickel Bier   

        review_palate  review_taste  \
651565            1.5           1.5   
659293            4.0           3.5   
659299            4.0           4.0   
659300            3.5           3.0   
659301            4.0           3.5   
659302            3.0           3.5   
659303            2.0           3.0   
659304            4.0           4.5   
659305            4.0           4.0   

                                                beer_name  beer_abv  \
651565  Engel Tyrolian Bräu WRONG BREWERY SEE SCHWABIS...       5.0   
659293   Engel Bock Dunkel WRONG BREWERY SEE CRAILSHEIMER       7.2   
659299          Engel Gold WRONG BREWERY SEE CRAILSHEIMER       5.4   
659300      Engel Landbier WRONG BREWERY SEE CRAILSHEIMER       4.8   
659301   Engel Keller Hell WRONG BREWERY SEE CRAILSHEIMER       5.4   
659302    Engel Aloisius - WRONG BREWERY SEE CRAILSHEIMER       5.9   
659303  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659304  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659305  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   

        beer_beerid  
651565        67503  
659293        63658  
659299        63215  
659300        63557  
659301        63256  
659302        63459  
659303        63324  
659304        63324  
659305        63324

Doesn't look like we have any brewery_id here however it's clean from the name of the beers and a little google search later, the brewery in this case is Engel located in Crailshaim, Germany.

In [9]:
df_ba.loc[df_ba['brewery_id'] == 1193, ['brewery_name']] = 'Engel' 

# Now let's check that worked to replace our brewery name

In [10]:
df_ba[df_ba['brewery_id'] == 1193]

# Looks like that worked, let's check out the other value

brewery_id brewery_name  review_time  review_overall  review_aroma  \
651565        1193        Engel   1301022066             2.0           2.5   
659293        1193        Engel   1290107698             4.0           4.5   
659299        1193        Engel   1289077001             3.5           3.0   
659300        1193        Engel   1289851033             3.5           4.0   
659301        1193        Engel   1289158632             3.5           4.0   
659302        1193        Engel   1289590065             3.5           4.0   
659303        1193        Engel   1298078926             3.0           3.0   
659304        1193        Engel   1292423271             4.5           4.0   
659305        1193        Engel   1289329962             3.5           3.5   

        review_appearance review_profilename                  beer_style  \
651565                2.5            Knapp85                Vienna Lager   
659293                3.5             dqrull                        Bock   
659299                3.0             dqrull   Dortmunder / Export Lager   
659300                3.5             dqrull         Munich Helles Lager   
659301                4.0             dqrull  Keller Bier / Zwickel Bier   
659302                3.0             dqrull                Vienna Lager   
659303                3.0         Ochsenblut  Keller Bier / Zwickel Bier   
659304                4.0         Dentist666  Keller Bier / Zwickel Bier   
659305                4.0             dqrull  Keller Bier / Zwickel Bier   

        review_palate  review_taste  \
651565            1.5           1.5   
659293            4.0           3.5   
659299            4.0           4.0   
659300            3.5           3.0   
659301            4.0           3.5   
659302            3.0           3.5   
659303            2.0           3.0   
659304            4.0           4.5   
659305            4.0           4.0   

                                                beer_name  beer_abv  \
651565  Engel Tyrolian Bräu WRONG BREWERY SEE SCHWABIS...       5.0   
659293   Engel Bock Dunkel WRONG BREWERY SEE CRAILSHEIMER       7.2   
659299          Engel Gold WRONG BREWERY SEE CRAILSHEIMER       5.4   
659300      Engel Landbier WRONG BREWERY SEE CRAILSHEIMER       4.8   
659301   Engel Keller Hell WRONG BREWERY SEE CRAILSHEIMER       5.4   
659302    Engel Aloisius - WRONG BREWERY SEE CRAILSHEIMER       5.9   
659303  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659304  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   
659305  Engel Keller Dunkel  WRONG BREWERY SEE CRAILSH...       5.3   

        beer_beerid  
651565        67503  
659293        63658  
659299        63215  
659300        63557  
659301        63256  
659302        63459  
659303        63324  
659304        63324  
659305        63324

In [11]:
df_ba[df_ba['brewery_id'] == 27]

brewery_id brewery_name  review_time  review_overall  review_aroma  \
1391043          27          NaN   1020244856             3.5           3.5   
1391049          27          NaN   1143182956             2.5           2.0   
1391050          27          NaN   1117963232             1.5           1.5   
1391051          27          NaN   1323314674             4.5           4.5   
1391052          27          NaN   1320989774             3.5           4.0   
1391053          27          NaN   1320995408             5.0           4.5   

         review_appearance review_profilename              beer_style  \
1391043                3.5              Jason  American Adjunct Lager   
1391049                2.0            BeerBob    American Malt Liquor   
1391050                2.0      feloniousmonk    American Malt Liquor   
1391051                3.5             ccrida            American IPA   
1391052                3.5              Docer            American IPA   
1391053                4.0              Docer          American Stout   

         review_palate  review_taste               beer_name  beer_abv  \
1391043            4.0           3.5  Hard Hat American Beer       3.8   
1391049            2.0           2.0             Side Pocket       8.5   
1391050            1.5           1.5             Side Pocket       8.5   
1391051            4.5           4.5           Breakaway IPA       7.2   
1391052            4.0           4.0           Breakaway IPA       7.2   
1391053            4.5           4.5   Caboose Oatmeal Stout       7.0   

         beer_beerid  
1391043           60  
1391049         3394  
1391050         3394  
1391051        75135  
1391052        75135  
1391053        75137

Similar story for brewery_id 27, doesn't look like this value exists at all within our dataframe. Let's try a similar technique by looking at the beer names to see if a brewery can be determined.

In [12]:
# American Brewing Company

In [13]:
df_ba.loc[df_ba['brewery_id'] == 27, ['brewery_name']] = 'American Brewing Company' 

In [14]:
df_ba['brewery_name'].isna().sum()

0

Nice, we've now dealt with all values in the brewery_name columns. Let's tackle another.

In [15]:
df_ba.isna().sum()

brewery_id                0
brewery_name              0
review_time               0
review_overall            0
review_aroma              0
review_appearance         0
review_profilename      348
beer_style                0
review_palate             0
review_taste              0
beer_name                 0
beer_abv              67785
beer_beerid               0
dtype: int64

Review_profilename is an interesting column. It's the name of the user who submitted the review and since we don't have a scoring system associated with each user, I'm just going to drop the column and assume that all beers are rated equally and fairly.

In [16]:
df_ba.drop(['review_profilename'], inplace=True, axis=1)

# Let us check to make sure this was dropped correctly.

In [17]:
df_ba.isna().sum()

# nice, only beer_abv to deal with

brewery_id               0
brewery_name             0
review_time              0
review_overall           0
review_aroma             0
review_appearance        0
beer_style               0
review_palate            0
review_taste             0
beer_name                0
beer_abv             67785
beer_beerid              0
dtype: int64

This is a solid chunk of our data. I think the best approach would be to take the mean beer_abv for each beer_style and then apply that to our data. It's not a fool proof approach however it certaily would give us a good indication of the ABV so we can use that data going forward. There certainly will be some outliers here however beer style is a generally good way of designating ABV as different styles have different ABVs.

In [18]:
beer_styles = list(df_ba['beer_style'].unique())
beer_styles[:5]

# Cool, looks like we have a unique list of different beers.

['Hefeweizen',
 'English Strong Ale',
 'Foreign / Export Stout',
 'German Pilsener',
 'American Double / Imperial IPA']

In [19]:
beer_abv = {}

for style in beer_styles:
        func_df = df_ba[df_ba['beer_style'] == style]
        mean = round(func_df['beer_abv'].mean(),1)
        beer_abv.update({style : mean})
                

In [20]:
beer_abv

# want to float these to only 1 decimal, updated function above.

{'Hefeweizen': 5.3,
 'English Strong Ale': 7.3,
 'Foreign / Export Stout': 7.7,
 'German Pilsener': 5.0,
 'American Double / Imperial IPA': 9.4,
 'Herbed / Spiced Beer': 6.6,
 'Light Lager': 4.0,
 'Oatmeal Stout': 5.8,
 'American Pale Lager': 5.0,
 'Rauchbier': 5.7,
 'American Pale Ale (APA)': 5.5,
 'American Porter': 6.2,
 'Belgian Strong Dark Ale': 9.4,
 'American IPA': 6.6,
 'American Stout': 6.3,
 'Russian Imperial Stout': 10.0,
 'American Amber / Red Ale': 6.0,
 'American Strong Ale': 9.8,
 'Märzen / Oktoberfest': 5.9,
 'American Adjunct Lager': 4.9,
 'American Blonde Ale': 5.0,
 'Euro Pale Lager': 5.1,
 'English Brown Ale': 5.2,
 'Scotch Ale / Wee Heavy': 8.2,
 'Fruit / Vegetable Beer': 5.8,
 'American Double / Imperial Stout': 10.6,
 'Belgian Pale Ale': 6.1,
 'English Bitter': 4.3,
 'English Porter': 5.7,
 'Irish Dry Stout': 4.9,
 'American Barleywine': 10.7,
 'Belgian Strong Pale Ale': 8.7,
 'Doppelbock': 8.3,
 'Maibock / Helles Bock': 6.8,
 'Pumpkin Ale': 6.6,
 'Dortmunder / E

In [21]:
beer_abv['Gose']

4.7

In [22]:
df_ba.isna().sum()

brewery_id               0
brewery_name             0
review_time              0
review_overall           0
review_aroma             0
review_appearance        0
beer_style               0
review_palate            0
review_taste             0
beer_name                0
beer_abv             67785
beer_beerid              0
dtype: int64

In [23]:
for beer in beer_abv:
    df_ba['beer_abv'] = df_ba['beer_abv'].fillna(beer_abv[beer])

In [24]:
print(df_ba.info())
df_ba.isna().sum()

# This dataset is very clean now.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 12 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   brewery_id         1586614 non-null  int64  
 1   brewery_name       1586614 non-null  object 
 2   review_time        1586614 non-null  int64  
 3   review_overall     1586614 non-null  float64
 4   review_aroma       1586614 non-null  float64
 5   review_appearance  1586614 non-null  float64
 6   beer_style         1586614 non-null  object 
 7   review_palate      1586614 non-null  float64
 8   review_taste       1586614 non-null  float64
 9   beer_name          1586614 non-null  object 
 10  beer_abv           1586614 non-null  float64
 11  beer_beerid        1586614 non-null  int64  
dtypes: float64(6), int64(3), object(3)
memory usage: 145.3+ MB
None


brewery_id           0
brewery_name         0
review_time          0
review_overall       0
review_aroma         0
review_appearance    0
beer_style           0
review_palate        0
review_taste         0
beer_name            0
beer_abv             0
beer_beerid          0
dtype: int64